### PR TITLE
Fix dtype casting in swinv2 and swinv2sr to allow non-FP32 inference

### PR DESCRIPTION
**Pull request by aliencaocao**
# What does this PR do?
The current implementation uses `.float()` in https://github.com/huggingface/transformers/blob/0f67ba1d741d65b07d549daf4ee157609ce4f9c1/src/transformers/models/swin2sr/modeling_swin2sr.py#L286-L287 which causes subsequent `relative_coords_table` to be always in `torch.float32`, not respecting whatever precision the other weights might be, e.g. `torch.float16`.

This PR adds a cast to the same `dtype` as the `continuous_position_bias_mlp` layer since `relative_coords_table` is being passed directly into the layer at https://github.com/huggingface/transformers/blob/0f67ba1d741d65b07d549daf4ee157609ce4f9c1/src/transformers/models/swin2sr/modeling_swin2sr.py#L349

Same issue & fix for swinv2

Prerequisite for https://github.com/huggingface/transformers/pull/31342 image to image pipeline FP16 test to pass.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@amyeroberts

### CODE REVIEW

File: src/transformers/models/swin2sr/modeling_swin2sr.py [hunk 1]
```diff
@@ -301,6 +301,9 @@ def __init__(self, config, dim, num_heads, window_size, pretrained_window_size=[
         relative_coords_table = (
             torch.sign(relative_coords_table) * torch.log2(torch.abs(relative_coords_table) + 1.0) / math.log2(8)
         )
+        relative_coords_table = relative_coords_table.to(
+            next(self.continuous_position_bias_mlp.parameters()).dtype
+        )  # set to same dtype as mlp weight
```
Comment (amyeroberts) [inline src/transformers/models/swin2sr/modeling_swin2sr.py:306]: nit - library convention is to place comments above to avoid line splitting (I realise this is somewhat violated above)

```suggestion
        # set to same dtype as mlp weight
        relative_coords_table = relative_coords_table.to(next(self.continuous_position_bias_mlp.parameters()).dtype)
```

File: src/transformers/models/swinv2/modeling_swinv2.py [hunk 2]
```diff
@@ -492,6 +492,9 @@ def __init__(self, config, dim, num_heads, window_size, pretrained_window_size=[
         relative_coords_table = (
             torch.sign(relative_coords_table) * torch.log2(torch.abs(relative_coords_table) + 1.0) / math.log2(8)
         )
+        relative_coords_table = relative_coords_table.to(
+            next(self.continuous_position_bias_mlp.parameters()).dtype
+        )  # set to same dtype as mlp weight
```
Comment (amyeroberts) [inline src/transformers/models/swinv2/modeling_swinv2.py:497]:
```suggestion
        # set to same dtype as mlp weight
        relative_coords_table = relative_coords_table.to(next(self.continuous_position_bias_mlp.parameters()).dtype)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing! 

Could you also add some small tests like this one for [vit](https://github.com/huggingface/transformers/blob/0f67ba1d741d65b07d549daf4ee157609ce4f9c1/tests/models/vit/test_modeling_vit.py#L321)?

**Comment by aliencaocao** [comment]
Added them, they pass locally

**Comment by amyeroberts** [comment]
@aliencaocao Great! Could you push an empty commit with the message: `[run_slow] swin2sr, swinv2`. I trust the tests are passing locally, but because of differences that can creep in because of hardware and env set-up, the logits can still be slightly different. So let's make sure the numbers match what's going to be running on the CI : )

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing this!

**Comment by aliencaocao** [comment]
@amyeroberts need your approval for slow tests

**Comment by aliencaocao** [comment]
ugh it seems the specific gpu used indeed has numerical difference...
I ran the tests and got logit using RTX 3080Ti, torch2.3.1+cu121, nvidia 555.95 driver, Windows 10

**Comment by amyeroberts** [comment]
@aliencaocao As the change is simple, and look OK, updating the tests to use the results from the CI runs should be OK

**Comment by aliencaocao** [comment]
How do I get the CI outputs?

**Comment by amyeroberts** [comment]
@aliencaocao Good question! Indeed, they're not part of the console output. Let me see if I can ssh in

**Comment by amyeroberts** [comment]
@aliencaocao Running on the runners, I get the following logits

swin2sr
```
tensor([[0.5454, 0.5542, 0.5640],
        [0.5518, 0.5562, 0.5649],
        [0.5391, 0.5425, 0.5620]], device='cuda:0', dtype=torch.float16)
```

swinv2
```
tensor([-0.3938, -0.4290,  0.0020], device='cuda:0', dtype=torch.float16)
```

**Comment by aliencaocao** [comment]
Thanks, triggered again

**Comment by amyeroberts** [comment]
@aliencaocao Thanks! All looks good - we can merge 🤗
